src/backend/distributed/commands/statistics.c

```diff
@@ -0,0 +1,234 @@
+/*-------------------------------------------------------------------------
+ *
+ * statistics.c
+ *    Commands for STATISTICS statements.
+ *
+ *    We currently support replicating statistics definitions on the
+ *    coordinator in all the worker nodes in the form of
+ *
+ *    CREATE STATISTICS ... queries.
+ *
+ * Copyright (c) Citus Data, Inc.
+ *
+ *-------------------------------------------------------------------------
+ */
+
+#include "postgres.h"
+
+#include "access/genam.h"
+#include "access/htup_details.h"
+#include "access/xact.h"
+#include "catalog/namespace.h"
+#include "catalog/pg_namespace.h"
+#include "catalog/pg_statistic_ext.h"
+#include "catalog/pg_type.h"
+#include "distributed/commands/utility_hook.h"
+#include "distributed/commands.h"
+#include "distributed/deparse_shard_query.h"
+#include "distributed/deparser.h"
+#include "distributed/listutils.h"
+#include "distributed/metadata_sync.h"
+#include "distributed/multi_executor.h"
+#include "distributed/namespace_utils.h"
+#include "distributed/relation_access_tracking.h"
+#include "distributed/resource_lock.h"
+#include "distributed/worker_transaction.h"
+#include "utils/builtins.h"
+#include "utils/fmgroids.h"
+#include "utils/lsyscache.h"
+#include "utils/ruleutils.h"
+#include "utils/syscache.h"
+
+static List * GetExplicitStatisticsIdList(Oid relationId);
+static void EnsureSequentialModeForStatisticsDDL(void);
+
+
+/*
+ * PreprocessCreateStatisticsStmt is called during the planning phase for
+ * CREATE STATISTICS.
+ */
+List *
+PreprocessCreateStatisticsStmt(Node *node, const char *queryString)
+{
+	CreateStatsStmt *stmt = castNode(CreateStatsStmt, node);
+
+	DDLJob *ddlJob = palloc0(sizeof(DDLJob));
+	RangeVar *relation = (RangeVar *) linitial(stmt->relations);
+	Oid relationId = RangeVarGetRelid(relation, AccessExclusiveLock, false);
```

Comment:
we should make sure to use the right lock type here to avoid blocking reads

CreateStatistics in statscmds.c only uses ShareUpdateExclusiveLock

Reply:
thank you! fixed